### DISCUSSION TITLE
Handling Inter-Occasion Variability (IOV) and Dynamic ETA Assignment in mapbayr

### DISCUSSION THREAD

**Discussion by marianklose** [FelicienLL/mapbayr#212] [General]
Thanks @FelicienLL for developing this neat package!

I have a question regarding the implementation of IOV. I assume that I need to define for each occasion an additional ETA(n) and ETAn to use it within mapbayr, correct? Is there by chance any example code available for handling IOV?

Additionally, I am currently exploring the possibility of dynamically determining the number of cycles (occasions) from R. I was wondering if alternative approaches like the one mentioned in the [mrgsolve issue #408](https://github.com/metrumresearchgroup/mrgsolve/issues/408) work with mapbayr? Or do I need to hardcode n dummy occasions and then estimate all additional ETAs no matter if I have let's say 1 or 10 occasions. 

Thanks for your help!

**Comment by FelicienLL** [comment]
Hi Marian 
Thanks for your comment, 

Your implementation of IOV is correct, it is the same as in NONMEM. In the validation paper we coded several models that include IOV, all the code is available in a dedicated repo here: https://github.com/FelicienLL/mapbayr-CPTPSP-2021

I do not have a straightforward method to dynamically determine IOV. The method you mentionned is a nice workaround for simulations, but cannot work with mapbayr, and never will: conceptually, mrgsolve assumes a single (non-time varying) vector of ETA values for each individual, so I want to stick to this. 

However, in `mapbayest()` you have a `select_eta` argument to decide which ETA you want to estimate (and let the others to ZERO). I don't know if it can fit to your problem but one can imagine that you hard code a lot of occasions and ETAs in the model code, and that you programmatically control the one you estimate with `select_eta`? I never did something like this (i.e. up to 10 occasions) but managing useless ETAs in the context of IOV is what I had in mind when developping the `select_eta` feature. 

Let me know if it worked for you and if I can help in any way.

Best regards

Félicien

**Reply by marianklose** [reply]
Hi Félicien,

thanks for the clarification. I will try out the proposed solution with n hard-coded 'dummy ETAs' + controlling the ETAs to be estimated via the `select_eta` parameter. Happy to share the code in case this works.

Best,
Marian

**Comment by marianklose** [comment]
For those who are interested in a possible implementation of this as mentioned by @FelicienLL:

## Define Model

We use the mcode function to read in the model code. These are the changes to the example from the documentation:

- OCC is passed from the dataset in R so we define OCC as a covariate in *$PARAM*
- for each IOV term we have a
  -  `ETAn: 0 : IOV on CL (OCC = n+2)` statement in the *$PARAM*
  -  `ETA_IOVn : 0.1 : IOV on CL (OCC = n+2)` statement in *$OMEGA*
  - `if (OCC == n) {IOV = ETA_IOVn+2;  IOV_MB = ETAn+2;)}` statement in *$MAIN*
- The IOV term is then added to CL (in this case) via `double CL = TVCL * exp(ETA1 + ETA_CL + IOV_MB + IOV) * pow(BW / 70, 1.2);`

Here is the updated example model code:

```
# define model code
code <- '
$PARAM @annotated
TVCL:  0.9 : Clearance
TVV1: 10.0 : Central volume
V2  : 10.0 : Peripheral volume of distribution
Q   :  1.0 : Intercompartmental clearance

ETA1: 0 : Clearance (L/h)
ETA2: 0 : Central volume (L)
ETA3: 0 : IOV on CL (OCC = 1)
ETA4: 0 : IOV on CL (OCC = 2)
ETA5: 0 : IOV on CL (OCC = 3)
ETA6: 0 : IOV on CL (OCC = 4)
ETA7: 0 : IOV on CL (OCC = 5)
ETA8: 0 : IOV on CL (OCC = 6)
ETA9: 0 : IOV on CL (OCC = 7)
ETA10: 0 : IOV on CL (OCC = 8)
ETA11: 0 : IOV on CL (OCC = 9)
ETA12: 0 : IOV on CL (OCC = 10)


$PARAM @annotated @covariates
BW  : 70   : Body weight (kg)
OCC : -99  : Occasion (will be passed by dataset)

$OMEGA @annotated
ETA_CL : 0.3 : IIV on CL
ETA_V1 : 0.3 : IIV on V1
ETA_IOV3 : 0.1 : IOV on CL (OCC = 1)
ETA_IOV4 : 0.1 : IOV on CL (OCC = 2)
ETA_IOV5 : 0.1 : IOV on CL (OCC = 3)
ETA_IOV6 : 0.1 : IOV on CL (OCC = 4)
ETA_IOV7 : 0.1 : IOV on CL (OCC = 5)
ETA_IOV8 : 0.1 : IOV on CL (OCC = 6)
ETA_IOV9 : 0.1 : IOV on CL (OCC = 7)
ETA_IOV10 : 0.1 : IOV on CL (OCC = 8)
ETA_IOV11 : 0.1 : IOV on CL (OCC = 9)
ETA_IOV12 : 0.1 : IOV on CL (OCC = 10)


$SIGMA
0.05 // proportional
0.1 // additive

$CMT @annotated
CENT  : Central compartment (mg/L)[ADM, OBS]
PERIPH: Peripheral compartment ()

$TABLE
double DV = (CENT/V1) *(1 + EPS(1)) + EPS(2);

$MAIN
// inter-occasion variablity
double IOV = 0;
double IOV_MB = 0;
if (OCC == 1) {
  IOV = ETA_IOV3;
  IOV_MB = ETA3;
} else if (OCC == 2) {
  IOV = ETA_IOV4;
  IOV_MB = ETA4;
} else if (OCC == 3) {
  IOV = ETA_IOV5;
  IOV_MB = ETA5;
} else if (OCC == 4) {
  IOV = ETA_IOV6;
  IOV_MB = ETA6;
} else if (OCC == 5) {
  IOV = ETA_IOV7;
  IOV_MB = ETA7;
} else if (OCC == 6) {
  IOV = ETA_IOV8;
  IOV_MB = ETA8;
} else if (OCC == 7) {
  IOV = ETA_IOV9;
  IOV_MB = ETA9;
} else if (OCC == 8) {
  IOV = ETA_IOV10;
  IOV_MB = ETA10;
} else if (OCC == 9) {
  IOV = ETA_IOV11;
  IOV_MB = ETA11;
} else if (OCC == 10) {
  IOV = ETA_IOV12;
  IOV_MB = ETA12;
}

double CL = TVCL * exp(ETA1 + ETA_CL + IOV_MB + IOV) * pow(BW / 70, 1.2) ;
double V1 = TVV1 * exp(ETA2 + ETA_V1) ;
double K12 = Q / V1  ;
double K21 = Q / V2  ;
double K10 = CL / V1 ;

$ODE
dxdt_CENT   =  K21 * PERIPH - (K10 + K12) * CENT ;
dxdt_PERIPH =  K12 * CENT - K21 * PERIPH ;

$CAPTURE DV CL
'

# build model 
my_model <- mcode("Example_model", code)
```


## Define Data

To define a dummy dataset to test the implementation I have used the my_data data from the docs and modified it. `OCC` will be passed as a covariate through this dataframe. With the function we create n occasions and we always add a bit of noise to `DV so we get different IOV terms for different cycles.

```
# Sample data frame
my_data <- data.frame(
  ID = 1,
  time = c(0,6,15,24),
  evid = c(1, rep(0,3)),
  cmt = 1,
  amt = c(100, rep(0,3)),
  rate = c(20, rep(0,3)),
  DV = c(NA, 3.9, 2.5, 1.1),
  mdv = c(1,0,0,1),
  BW = 90,
  OCC = 1
)

# Function to generate additional occasions
generate_occasions <- function(df, n_occasions) {
  # Create an empty list to hold the new data frames
  dfs_list <- vector("list", n_occasions)
  
  # Loop over the number of occasions to create new data frames
  for (i in 1:n_occasions) {
    new_df <- df %>%
      mutate(
        OCC = OCC + i,          # Increment the OCC by i
        time = time + 24 * i    # Increment the time by 24 * i
      ) %>%
      mutate(
        DV = ifelse(is.na(DV), NA, DV + runif(n(), -0.8, 0.8))  # Add random noise to DV
      )
    dfs_list[[i]] <- new_df
  }
  
  # Combine the original and new data frames
  final_df <- bind_rows(df, do.call(bind_rows, dfs_list))
  
  return(final_df)
}

# Generate 10 additional occasions
new_data <- generate_occasions(my_data, 9)


# Print the new data
print(new_data)
```

Output:

<html>
<body>


ID | time | evid | cmt | amt | rate | DV | mdv | BW | OCC
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 0 | 1 | 1 | 100 | 20 | NA | 1 | 90 | 1
1 | 6 | 0 | 1 | 0 | 0 | 3.9000000 | 0 | 90 | 1
1 | 15 | 0 | 1 | 0 | 0 | 2.5000000 | 0 | 90 | 1
1 | 24 | 0 | 1 | 0 | 0 | 1.1000000 | 1 | 90 | 1
1 | 24 | 1 | 1 | 100 | 20 | NA | 1 | 90 | 2
1 | 30 | 0 | 1 | 0 | 0 | 4.0048334 | 0 | 90 | 2
1 | 39 | 0 | 1 | 0 | 0 | 2.4066416 | 0 | 90 | 2
1 | 48 | 0 | 1 | 0 | 0 | 1.7442728 | 1 | 90 | 2
1 | 48 | 1 | 1 | 100 | 20 | NA | 1 | 90 | 3
... | ... | ... | ... | ... | ... | ... | ... | ... | ...
1 | 216 | 1 | 1 | 100 | 20 | NA | 1 | 90 | 10
1 | 222 | 0 | 1 | 0 | 0 | 3.3403892 | 0 | 90 | 10
1 | 231 | 0 | 1 | 0 | 0 | 2.6628701 | 0 | 90 | 10
1 | 240 | 0 | 1 | 0 | 0 | 0.6247739 | 1 | 90 | 10


</body>
</html>


## Estimation: 10 Occasions

First we can test it with all 10 Occasions in the dataset:

```
# estimate
my_est <- mapbayest(my_model, data = new_data)

# show estimates
print(my_est)
```

Model: Example_model 
ID : 1 individual(s).
OBS: 20 observation(s).
**ETA: 12 parameter(s) to estimate.**

Estimates: 
  ID      ETA1     ETA2       ETA3       ETA4        ETA5       ETA6      ETA7
1  1 0.1404683 0.873838 -0.2643984 0.01137879 -0.08553752 0.04768208 0.1803825
         ETA8         ETA9     ETA10     ETA11      ETA12
1 -0.05416061 -0.005971572 0.1174665 0.0473555 0.05263258


![image](https://github.com/FelicienLL/mapbayr/assets/132135131/c2a4b2b5-09a8-41b1-b8ee-9c0c1a34fd79)

We can see that we have different CL terms per occasion depending on the OCC due to the added IOV. mapbayr estimated 12 EBEs (2 IIVs and 10 IOVs) according to the `ETA: 12 parameter(s) to estimate` output line from `my_est`.


## Estimation: 3 Occasions

Now only for 3 occasions. We just subset the dataframe via `dplyr::filter(OCC <= 3)` and then pass it to `mapbayest()`. The `select_eta` parameter is now set to 2+nOCC (2 for IIV on CL and V1 and nOCC for the IOV terms on CL).

```
# filter data
new_data_filtered <- new_data %>%
  dplyr::filter(OCC <= 3)

# estimate
my_est <- mapbayest(my_model, data = new_data_filtered, select_eta = 1:(max(new_data_filtered$OCC)+2))

# show estimates
print(my_est)
```

Model: Example_model 
ID : 1 individual(s).
OBS: 6 observation(s).
ETA: 5 parameter(s) to estimate.

Estimates: 
  ID      ETA1      ETA2       ETA3      ETA4      ETA5
1  1 0.1182608 0.6607996 -0.1887232 0.2143864 0.0137581



![image](https://github.com/FelicienLL/mapbayr/assets/132135131/d00eb1e2-9ea8-4388-a20e-ec1637fbfd9e)

The `ETA: 5 parameter(s) to estimate.` output indicates that it worked and we avoided un-needed estimation of the remaining 7 IOV terms. 

Overall it is surely a bit messy to set up since you need to add 1 ETAn and 1 ETA(n) per occasion but once it is set up in the mrgsolve code you can dynamically decide from R how many EBEs you want to estimate to save some time / computational cost.
